### PR TITLE
New selections features

### LINKED ISSUE #73: Selections within a selection
It is for now possible to select atoms, bonds, angles or dihedrals on the base of their components' name, type, coordinates...
It could be interesting to select atoms based on their bonded neighbours. We could create a selection inside a selection, for example:
"atoms: name O and bonded to {name H and x<5}"

### PR DESCRIPTION
**Pull request by Luthaf**
This adds a few more features to selections:

- mathematical functions depending on multiple atoms `distance(#1, #2)`, `angle(#1, #2, #3)`, `dihedral(#1, #2, #3, #4)`, and `out_of_plane(#1, #2, #3, #4)`
- topology constrains: `bonded(#1, #2)`, `is_angle(#1, #2, #3)`, `is_dihedral(#1, #2, #3, #4)`, and
      `improper(#1, #2, #3, #4)`
- sub-selections in topology constrains: `bonded(#1, name O)`, `is_angle(name H, name C, #2)`, ...
- implement raw-identifiers to be used as string values (7ef38e3)

@frodofine would you have some time to give this a look ? I would specially appreciate any improvement to the documentation.

Fix #73

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,9 +7,24 @@ This project adheres to [Semantic Versioning](http://semver.org/).
...
+      expressions: `distance(#1, #2)`, `angle(#1, #2, #3)`, `dihedral(#1, #2,
+      #3, #4)`, and `out_of_plane(#1, #2, #3, #4)` are supported;
+    * it is now possible to add constrains on the topology of the system:
+      `bonded(#1, #2)`, `is_angle(#1, #2, #3)`, `is_dihedral(#1, #2, #3, #4)`,
+      and `improper(#1, #2, #3, #4)` are supported;
+    * the topology constrains support sub-selections: instead of checking is
+      `#1` and `#2` are bonded, one can check if `#1` is bonded to any atom
+      matching a selection, for example `name O` with `bonded(#1, name O)`.
+    * When using numbers as atomic names/types, they must now be inside double
+      quotes (`name "45"`). This also allow for more exotic atomic names
```
Comment (frodofine) [inline CHANGELOG.md]: This allow**s** for exotic atomic names, e.g. (`name "名"`).

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,9 +7,24 @@ This project adheres to [Semantic Versioning](http://semver.org/).
...
+  PDB format.
+* The selection engine has been rewritten to add support for more complex
+  selections:
+    * it is now possible to use mathematical expressions in selections such as
+      `x^2 - y ^2 < sqrt(z^2 + 25)`;
+    * it is now possible to access geometrical properties in such mathematical
+      expressions: `distance(#1, #2)`, `angle(#1, #2, #3)`, `dihedral(#1, #2,
+      #3, #4)`, and `out_of_plane(#1, #2, #3, #4)` are supported;
+    * it is now possible to add constrains on the topology of the system:
+      `bonded(#1, #2)`, `is_angle(#1, #2, #3)`, `is_dihedral(#1, #2, #3, #4)`,
```
Comment (frodofine) [inline CHANGELOG.md]: I'm confused. Why is there an is_ in front of some of the function names?

File: doc/src/selections.rst [hunk 3]
```diff
@@ -54,11 +53,36 @@ false and false to true. ``name(#1) == H and not x(#1) < 5.0`` and ``(z(#2) < 45
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Here is the list of currently implemented properties (either string or numeric
-properties). Additional properties ideas are welcome!
+Here is the list of currently implemented selectors. Additional ideas are welcome!
+
+Boolean selectors
+-----------------
+
+- ``all``: always matches the atom and return ``true``;
```
Comment (frodofine) [inline doc/src/selections.rst]: always matches any atom and returns

File: doc/src/selections.rst [hunk 4]
```diff
@@ -54,11 +53,36 @@ false and false to true. ``name(#1) == H and not x(#1) < 5.0`` and ``(z(#2) < 45
...
+- ``none``: that matches no atom and always return ``false``;
```
Comment (frodofine) [inline doc/src/selections.rst]: never matches an atom and returns

File: doc/src/selections.rst [hunk 5]
```diff
@@ -54,11 +53,36 @@ false and false to true. ``name(#1) == H and not x(#1) < 5.0`` and ``(z(#2) < 45
...
+- ``bonded(i, j)``: check if atoms i and j are bonded together. If i and j
```
Comment (frodofine) [inline doc/src/selections.rst]: Maybe rename is_bonded as to be consistent with the other functions?

File: doc/src/selections.rst [hunk 6]
```diff
@@ -73,15 +97,34 @@ String properties
...
+
+- ``distance(i, j)``: gives the distance in Ångströms between atoms i and j,
+  accounting for periodic boundary conditions.
+- ``angle(i, j, k)``: gives the angle between atoms i, j and k in radians,
+  accounting for periodic boundary conditions. The atoms do not need to be
+  bonded together.
+- ``dihedral(i, j, k, m)``: gives the dihedral angle between atoms i, j, k and m
+  in radians, accounting for periodic boundary conditions. The atoms do not need
+  to be bonded together.
+- ``out_of_plane(i, j, k, m)``: gives the distance in Ångströms between the
```
Comment (frodofine) [inline doc/src/selections.rst]: Maybe rename out_of_plane to improper to match the is_improper method.

File: CHANGELOG.md (hunk 2 above)
Comment (Luthaf) [reply CHANGELOG.md]: I'll clarify it in the doc, but there are two different family of functions. The ones with the `is_` prefix returns a boolean value and check the chemical topology, and the ones without the prefix returns the value of the geometrical quantity (angle, distance, dihedral, ...)

File: doc/src/selections.rst (hunk 6 above)
Comment (Luthaf) [reply doc/src/selections.rst]: As I understand it, improper dihedral angles can be used in two different ways when computing potentials: either as dihedral angles (the angle between ij and the jkm plane), or as an out of plane distance (distance between j and the ikm plane). This is why the functions are named this way. 

```
    i
    |
    j
  /   \
k       m
```

Does it makes sense ? I am not really used to work with impropers.

File: doc/src/selections.rst (hunk 5 above)
Comment (Luthaf) [reply doc/src/selections.rst]: Done

**Pull request merged**